File: leetcode/algorithms/hash-table/valid-anagram.java
/*
    leetcode
    category: "Algorithms"
    problem title: "Valid Anagram"
*/

class Solution {
    private HashMap<Character,Integer> getCharCounts(String str){
        HashMap<Character,Integer> charCounts = new HashMap<Character,Integer>();
        for(int i=0;i<str.length();++i){
            charCounts.put(str.charAt(i),charCounts.getOrDefault(str.charAt(i),0)+1);
        }
        return charCounts;
    }
    public boolean isAnagram(String s, String t) {
        HashMap<Character,Integer> tChars = getCharCounts(t);
        HashMap<Character,Integer> sChars = getCharCounts(s);
        
        boolean isAnagram=true;
        if(tChars.size()==sChars.size()){
            for(Map.Entry<Character,Integer> entry : tChars.entrySet()){
                Integer sCount = sChars.getOrDefault(entry.getKey(),null);
                if(sCount==null || sCount.intValue()!=entry.getValue()){
                    isAnagram=false;
                    break;
                }
            }
        } else {
            isAnagram = false;
        }
        
        return isAnagram;
        
    }
}

File: leetcode/algorithms/hash-table/valid-anagram.js
/*
    leetcode
    category: "Algorithms"
    problem title: "Valid Anagram"
*/

/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */


const getCharMap = (str) => {
    const charMap = new Map();
    for(let i=0;i<str.length;++i){
        charMap.set(str[i],(charMap.get(str[i]) || 0)+1);
    }
    return charMap;
}

var isAnagram = function(s, t) {
    const sMap = getCharMap(s);
    const tMap = getCharMap(t);
    
    let isValidAnagram = true;
    if(sMap.size===tMap.size){
        for(const [c,count] of sMap.entries()){
            if(!tMap.has(c) || tMap.get(c)!=count){
                isValidAnagram = false;
                break;
            }
        }
    } else {
        isValidAnagram = false;
    }
    
    return isValidAnagram;
    
};